From: Miguel Vadillo <vadillo@ti.com>
Date: Tue, 1 Mar 2011 16:59:17 +0000 (-0600)
Subject: syslink: ipu_pm: fix a warning in restore_ctx
X-Git-Tag: android-2.6.35-2.3-omap4.13.1~87
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=50823246c70df89edac559d56fe483ebaa2a6839

syslink: ipu_pm: fix a warning in restore_ctx

As part of the iommu close iommu_restore is called
but then as part of the detach messages are sending
to ipu and the restore_ctx is called and then iommu_restore
without checking if it was already restored causing a
warning.

Check if the iommu has not been restored and only then
restore it.

Change-Id: Ia5ff12a7a8d2e68f5d56ae26fd7985e058a37a2e
Reported-by: Subin Gangadharan <subin.kg@ti.com>
Signed-off-by: Miguel Vadillo <vadillo@ti.com>
---

Index: kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c
===================================================================
--- kernel.orig/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-05-10 20:14:06.000000000 +0900
+++ kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-05-10 20:14:12.000000000 +0900
@@ -2650,7 +2650,7 @@
 			iommu_restore_ctx(ducati_iommu);
 			_is_iommu_up = 1;
 		} else
-			pr_err("Not able to restore iommu");
+			pr_err("Not restoring iommu");
 
 		pr_info("Wakeup SYSM3\n");
 		retval = rproc_wakeup(sys_rproc);
@@ -2989,13 +2989,6 @@
 			/*
 			 * Restore iommu to allow process's iommu cleanup
 			 * after ipu_pm is shutdown
-			 */
-			if (ipu_pm_get_state(SYS_M3) & SYS_PROC_DOWN)
-				iommu_restore_ctx(ducati_iommu);
-			iommu_unregister_notifier(ducati_iommu,
-					&ipu_pm_notify_nb_iommu_ducati);
-			pr_debug("releasing ducati_iommu\n");
-			/*
 			 * FIXME: this need to be checked by the iommu driver
 			 * restore IOMMU since it is required the IOMMU
 			 * is up and running for reclaiming MMU entries
@@ -3004,6 +2997,9 @@
 				iommu_restore_ctx(ducati_iommu);
 				_is_iommu_up = 1;
 			}
+			iommu_unregister_notifier(ducati_iommu,
+					&ipu_pm_notify_nb_iommu_ducati);
+			pr_debug("%s ducati_iommu put\n", __func__);
 			iommu_put(ducati_iommu);
 			ducati_iommu = NULL;
 		}
